Fix for error cases in promql eval of StepInvariantExpr with unnecessary ParentExpr (Issue 16880)

tcp13equals2 opened the pull request:
This PR relates to https://github.com/prometheus/prometheus/issues/16880

During Mimir Query Engine regression testing, a number of edge case promql queries returned an error in the Prometheus Query Engine. 

The edge cases involved redundant () around expressions, resulting in the additional wrapping of expressions with ParentExprs. 

During the eval() processing of parser.StepInvariantExpr, the code paths related to returning early on literals & matrix selectors were not taken, resulting in an error.

It seems that most of the other parser types already have specific handling for removing redundant ParentExprs. 

The suggested fix is to remove root level ParentExpr's early in the PreprocessExpr().

Unit tests with an example query string causing an error have been included.

bboreham reviewed (changes requested):
Wait I didn’t realise how this changed since the last time I looked. Let me think about it some more.

tcp13equals2 commented:
@bboreham - thanks for reviewing and your comments. 

I've updated the comments on PreProcessExpr.

Note - from the previous PR I removed the unwrapParenExpr() in the eval() StepInvariantExpr type. I have not found a test case which requires the addition of this line, so I did not want to add a line which did not have a test which requires it. 

If you think it would be beneficial I can look at a future PR which seeks to remove the need for those other unwrapParenExpr() spread throughout the eval() type handling.

bboreham commented:
> The suggested fix is to remove root level ParentExpr's early in the PreprocessExpr().

Apologies, I did not pick up on this "root level" part until you emphasised it in the comments.
That means this change works for `("foo")` but not for `(("foo"))`, right?
Should we accept that limitation?

bboreham commented:
> That means this change works for ("foo") but not for (("foo")), right?

Wrong. Turns out `unwrapParenExpr()` would unwrap any number of levels. 

Since I've made so many mistakes trying to review this short PR, I decided to take a much closer look.
I've always felt a bit suspicious about manipulating the AST during evaluation, so it seems better to do all the adding and removing during the preprocess phase.  I came up with #16896.
